### ISSUE TITLE
Feature Request: LDAP support for Linux Server

### ISSUE THREAD

**Issue by EULABreaker** [Ylianst/MeshCentral#165]
Enable linux hosted MC2 server to call authentication against LDAP. 

Example:
I have a Windows Server 2016 AD Server. I have setup mesh central on a Ubuntu 18.04 Server to keep the two systems separated. Being able to connect the authentication from MC2 to AD via LDAP would be awesome.

Based on this seems to be almost %100 node / js something like passport-ldapauth could be used to connect to the LDAP server.

Thanks

**Comment by Ylianst** [comment]
Agreed, that would be a good feature. Do you have a quick way I could setup a Ubuntu machine with LDAP as a test machine? I have never setup a AD server, let alone one on Linux.

**Comment by EULABreaker** [comment]
The AD Server I use is a windows active directory server. Im running the Mesh Central Server on a Ubuntu 18.04 Server. I am glad to help in anyway needed.

The quickest way I can think of would be to just install Turnkey Linux Domain Controller.

https://www.turnkeylinux.org/domain-controller

Also if needed I can setup a AD Server for you on a Test Environment and you can work off that one. PM me if you would like me to setup a test Environment.

Thanks

**Comment by Ylianst** [comment]
If you are offering to setup a dummy Linux with AD, I will take you up on it. If you can setup a public facing Linux with AD that has MeshCentral on it... yes, I will add AD support to it. Mail me the machine credentials and I will get it done.

**Comment by EULABreaker** [comment]
I will get it started tomorrow. Thank you

**Comment by D4V3M0NK** [comment]
For my LDAP environments, I use the 90 day free server downloads from [Microsoft](https://www.microsoft.com/en-us/evalcenter/evaluate-windows-server-2019?filetype=ISO) - versions back to 2012 are available. 

- Pop it into a VM (I use VirtualBox) 
- Snapshot it immediately. 
- Once the server is up and running, configure as required. You've got 90 days to test. (*)
- Once the 90 days are up, restore back to the snapshot and then bring it up again.
- Reconfigure, then you're good for another 90 days.

(*) I've just thought of this, but you also may be able to create a backup once it's configured that you could use after reconfiguring.

I'd be more than willing to assist in this area too, should you guys need it.

**Comment by EULABreaker** [comment]
Servers are setup and email is sent.

**Comment by Ylianst** [comment]
Got it.

**Comment by Ylianst** [comment]
Published MeshCentral v0.3.2-k that includes for the first time LDAP support. In the domain section of config.json, add these two values:

```
      "auth": "ldap",
      "ldapOptions": {
        "url": "ldap://1.2.3.4:389",
        "bindDN": "CN=meshcentral,...",
        "bindCredentials": "password",
        "searchBase": "DC=meshcentral,DC=local",
        "searchFilter": "(sAMAccountName={{username}})"
      }
```

MeshCentral uses [ldapauth-fork](https://www.npmjs.com/package/ldapauth-fork), so the options that need to be provided in the "ldapOptions" are the same as this library. Testing and feedback appreciated.

**Comment by TitovLab** [comment]
It works!
After logging in using LDAP, even if sAMAccountName is used, MeshCentral then uses the CN as the username. CN, especially in Active Directory, contains the name, surname and degree. Very often CN will contain not only English characters, which leads to two errors:
1 MeshCentral can't use non-English characters in username for issuing rights to Device Groups
2 If user with non-English CN creates a new Device Group, it appears in the interface only after reloading browser window
Given that LDAP is sometimes not Active Directory, it may make sense in the configuration to be able to set which LDAP attribute to use for username? For example, "LDAPusernameField": "sAMAccountName"?
Also, with the advent of LDAP, it makes sense by default to enable "No New Device Groups" for users or, more likely, to determine what rights the created users should be granted by default.
Thank you!

**Comment by Ylianst** [comment]
Your absolutely right. I need to make a bunch of changes to the server to fix places where I use the userid/name incorrectly. After that, i should be able to do this. One problem I will be having is that when you want to add a user to a device group for example, I will need to have some user lookup. But yes, your right about this and allowing that field to use.

**Comment by Ylianst** [comment]
Published MeshCentral v0.3.2-n with LDAP improvements. The new server can better handle the difference between the username and userid. You can also configure that field is used as display name and user identifier.

      "ldapUserName": "displayName",
      "ldapUserBinaryKey": "objectSid",
      "ldapOptions": {
        "url": "ldap://1.2.3.4:389",
        "bindDN": "CN=meshcentral,CN=Users,DC=meshcentral,DC=local",
        "bindCredentials": "password",
        "searchBase": "DC=meshcentral,DC=local",
        "searchFilter": "(sAMAccountName={{username}})"
      }

The one big problem now is that to add new user to a device group, you need to add the user using the userid. I will need to add a user lookup system. Testing of this version appreciated.

**Comment by Ylianst** [comment]
This request should be done and I will close it. However, I do need to implement user lookup to make this feature work correctly. That is already filed  #177. So I will close this one. Please re-open if needed.
